scene/gui/button.cpp

```diff
@@ -72,287 +129,266 @@ void Button::_notification(int p_what) {
+						case VERTICAL_ALIGNMENT_TOP: {
+							icon_ofs.y += style_margin_top;
+						} break;
+
+						case VERTICAL_ALIGNMENT_BOTTOM: {
+							icon_ofs.y = size.y - style_margin_bottom - icon_size.height;
+						} break;
+					}
 
-				if (vertical_icon_alignment == VERTICAL_ALIGNMENT_TOP) {
-					voffset = -(valign - icon_size.y) / 2;
-				}
-				if (vertical_icon_alignment == VERTICAL_ALIGNMENT_BOTTOM) {
-					voffset = (valign - icon_size.y) / 2 + text_buf->get_size().y;
+					Rect2 icon_region = Rect2(icon_ofs, icon_size);
+					draw_texture_rect(_icon, icon_region, false, icon_modulate_color);
 				}
 
-				if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_LEFT) {
-					icon_region = Rect2(style_offset + Point2(icon_ofs_region, voffset + Math::floor((valign - icon_size.y) * 0.5)), icon_size);
-				} else if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_CENTER) {
-					icon_region = Rect2(style_offset + Point2(icon_ofs_region + Math::floor((size.x - icon_size.x) * 0.5), voffset + Math::floor((valign - icon_size.y) * 0.5)), icon_size);
-				} else {
-					icon_region = Rect2(style_offset + Point2(icon_ofs_region + size.x - icon_size.x, voffset + Math::floor((valign - icon_size.y) * 0.5)), icon_size);
-				}
+				if (!xl_text.is_empty()) {
+					// Update the size after the icon is stripped. Stripping only when the icon alignments are not center.
+					if (icon_align_rtl_checked != HORIZONTAL_ALIGNMENT_CENTER) {
+						// Subtract the space's width occupied by icon and h_separation together.
+						drawable_size_remained.width -= icon_size.width + h_separation;
+					}
 
-				if (icon_region.size.width > 0) {
-					Rect2 icon_region_rounded = Rect2(icon_region.position.round(), icon_region.size.round());
-					draw_texture_rect(_icon, icon_region_rounded, false, color_icon);
+					if (vertical_icon_alignment != VERTICAL_ALIGNMENT_CENTER) {
+						drawable_size_remained.height -= icon_size.height;
+					}
 				}
 			}
 
-			Point2 icon_ofs = !_icon.is_null() ? Point2(icon_region.size.width + theme_cache.h_separation, 0) : Point2();
-			if (align_rtl_checked == HORIZONTAL_ALIGNMENT_CENTER && icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_CENTER) {
-				icon_ofs.x = 0.0;
-			}
-
-			int text_clip = size.width - style->get_minimum_size().width - icon_ofs.width;
-			if (_internal_margin[SIDE_LEFT] > 0) {
-				text_clip -= _internal_margin[SIDE_LEFT] + theme_cache.h_separation;
-			}
-			if (_internal_margin[SIDE_RIGHT] > 0) {
-				text_clip -= _internal_margin[SIDE_RIGHT] + theme_cache.h_separation;
-			}
-
-			text_buf->set_width(is_clipped ? text_clip : -1);
+			// Draw the text.
+			if (!xl_text.is_empty()) {
+				text_buf->set_alignment(align_rtl_checked);
 
-			int text_width = MAX(1, is_clipped ? MIN(text_clip, text_buf->get_size().x) : text_buf->get_size().x);
+				float text_buf_width = MAX(1.0f, drawable_size_remained.width); // The space's width filled by the text_buf.
+				text_buf->set_width(text_buf_width);
```

Comment:
Due to the conditions added in #80402, the calculated "drawable_size_remained.width" is less affected by "text_buf->get_size()". That is to say, the calculated results are more reliable.

If I understand "text_buf->set_width()" correctly, doing this will align correctly in the remaining space.

Reply:
Visible characters will be offset in the text‘s reserve space by `TextParagraph`. 

So we can later calculate the offset to this space from the left, as the required values are all determined.